internal/fs/hash.go

```diff
@@ -0,0 +1,145 @@
+package fs
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	pathSeparator = string(filepath.Separator)
+	skipModes     = os.ModeDevice | os.ModeNamedPipe | os.ModeSocket | os.ModeCharDevice
+)
+
+// HashFromNode returns a deterministic hash of the specified file system node,
+// performing a breadth-first traversal of directories. While the specified
+// prefix is joined with the pathname to walk the file system, the prefix string
+// is eliminated from the pathname of the nodes encounted when hashing the
+// pathnames.
+//
+// This function ignores any file system node named `vendor`, `.bzr`, `.git`,
+// `.hg`, and `.svn`, as these are typically used as Version Control System
+// (VCS) directories.
+//
+// Other than the `vendor` and VCS directories mentioned above, the calculated
+// hash includes the pathname to every discovered file system node, whether it
+// is an empty directory, a non-empty directory, empty file, non-empty file, or
+// symbolic link. If a symbolic link, the referent name is included. If a
+// non-empty file, the file's contents are incuded. If a non-empty directory,
+// the contents of the directory are included.
+//
+// While filepath.Walk could have been used, that standard library function
+// skips symbolic links, and for now, we want to hash the referent string of
+// symbolic links.
+func HashFromNode(prefix, pathname string) (hash string, err error) {
```

Comment:
I buy everything you suggest here, but after working with the code, and keeping in mind the fact that `Gopkg.lock` has string versions of digest sums in them, I have a hard time seeing this as `[]byte` rather than a `string`. I'm very curious what you are thinking about that I am not to suggest that `[]byte` will work better than `string` here.